pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,461 @@
+        Map<String, String> mergeConfigs = sortedMergeLevelConfigs.get(i).getValue();
+
+        // Skip scheduling if there's incomplete task for current mergeLevel
+        if (inCompleteMergeLevels.contains(mergeLevel)) {
+          LOGGER.info("Found incomplete task of merge level: {} for the same table: {}, Skipping task generation: {}",
+              mergeLevel, offlineTableName, taskType);
+          continue;
+        }
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+        long bufferMs =
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+
+        // Get watermark from MergeRollupTaskMetadata ZNode
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        long waterMarkMs =
+            getWatermarkMs(preSelectedSegments.first().getStartTimeMs(), bucketMs, mergeLevel, mergeRollupTaskMetadata);
+        long windowStartMs = waterMarkMs;
+        long windowEndMs = windowStartMs + bucketMs;
+
+        if (!isValidMergeWindowEndTime(windowEndMs, bufferMs, lowerMergeLevel, mergeRollupTaskMetadata)) {
+          LOGGER.info(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          continue;
+        }
+
+        // Find all segments overlapping with the merge window, if all overlapping segments are merged, bump up the target window
+        List<OfflineSegmentZKMetadata> selectedSegments = new ArrayList<>();
+        boolean hasUnMergedSegments = false;
+        boolean isValidMergeWindow = true;
+
+        // The for loop terminates in following cases:
+        // 1. Found unmerged segments in target merge window, need to bump up watermark if windowStartMs > watermarkMs,
+        //    will schedule tasks
+        // 2. All segments are merged in the merge window and we have loop through all segments, need to bump up
+        //    watermark to windowStartMs + bucketTimeMs, skip scheduling
+        // 3. Current watermark > endTimeMs of all segments, won't bump up watermark, skip scheduling
+        // 4. Merge window is invalid (windowEndMs > System.currentTimeMillis() - bufferMs || windowEndMs > waterMark of
+        //    the lower mergeLevel), may need to bump up watermark if windowStartMs > watermarkMs, skip scheduling
+        for (OfflineSegmentZKMetadata preSelectedSegment : preSelectedSegments) {
+          if (windowStartMs <= preSelectedSegment.getEndTimeMs() && preSelectedSegment.getStartTimeMs() < windowEndMs) {
+            // For segments overlapping with current window, add to the result list
+            selectedSegments.add(preSelectedSegment);
+            if (!isMergedSegment(preSelectedSegment, mergeLevel)) {
```

Comment:
What happens if we assign to merge the previously merged segment?

1st round:
s1, s2 -> merged_s3, merged_s4, merged_s5   # merged_s3 is for `07/01/2021`, merged_s4, merged_s5 are for `07/02/2021`.

2nd round:
merged_s4, merged_s5 -> merged_s6, merged_s7

====

Let's assume that we pre-selected `merged_s4, merged_s5` because the task is not scheduled yet.

When we reach this check, I think that `merged_s4,merged_s5` indicates that they are merged segments so we will have `hasUnMergedSegments = false` while those are not merged yet on the perspective of the current window.

Are we correctly handle this case?

Reply:
For this case, all segments in current window are merged, we should bump up the scheduling window. There are 2 cases:
1. there are un-merged segments after current window(e.g. s3 on `07/03/2021`). The for loop will enter the `else` condition on line `242`. It will bump up current window, and starts to pick s3.
2. there's no un-merged segment after current window. The function will go to line `276` to bump up the watermark, and skip scheduling on line `290`.
So I think it's handled.

Reply:
Added a `TODO` for future improvement that if there are many small segments which are marked as merged by previous merge cycle, we should merge them again rather than skipping.

Reply:
As discussed offline, I think that correctness-wise it's fine but we can improve the scheduler later by scheduling segments again.